gobblin-core/src/main/java/org/apache/gobblin/writer/partitioner/TimeBasedAvroWriterPartitioner.java

```diff
@@ -73,9 +74,25 @@ public long getRecordTimestamp(GenericRecord record) {
   /**
    *  Check if the partition column value is present and is a Long object. Otherwise, use current system time.
    */
-  private static long getRecordTimestamp(Optional<Object> writerPartitionColumnValue) {
-    return writerPartitionColumnValue.orNull() instanceof Long ? (Long) writerPartitionColumnValue.get()
-        : System.currentTimeMillis();
+  private long getRecordTimestamp(Optional<Object> writerPartitionColumnValue) {
+
+    // Default to current time
+    Long recordTimestamp = timeUnit.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+
+    if (writerPartitionColumnValue.isPresent()) {
+      Object val = writerPartitionColumnValue.get();
+      if (val instanceof Long) {
+        recordTimestamp = (Long) val;
+      } else {
+        try {
+          recordTimestamp = Long.parseLong(val.toString());
+        } catch (NumberFormatException e) {
+          // ignore
```

Comment:
This may mask errors and the behavior of parseable values being used and other cases getting the latest time may not be desired.

I'm not sure why the old code allows an invalid partition column to be specified, but this new code will add conversion overhead and may change the behavior.

So I think a config should be added to maintain the old behavior by default. And maybe there should also be a config to raise an error immediately on parsing errors to help detect bad data.

Reply:
I'll add a config to enable parsing string as long